File: 2020/day2/python/day2.py
#!/usr/bin/env python3
from os.path import dirname, realpath, join
import time, pytest

def transform_input(input_):
  # custom transform for the day
  count, letter, pw = input_.replace(':','').split()
  low, high = count.split('-')
  return int(low), int(high), letter, pw


def read_input(file_name = '../data/input.txt'):
  dir_path = dirname(realpath(__file__))
  with open(file_name, 'r') as f:
    input_ = [transform_input(x.strip()) for x in f.readlines()]

  return input_

def solve_part1(input_):
  n_pws = 0
  for low, high, letter, pw in input_:
    n_pws += 1 if low <= pw.count(letter) <= high else 0
  return n_pws
    

def solve_part2(input_):
  n_pws = 0
  for low, high, letter, pw in input_:
    n_pws += 1 if (pw[low-1] == letter) ^ (pw[high-1] == letter) else 0
  return n_pws


def main():
  
  input_ = read_input()

  t0 = time.time()
  part1 = solve_part1(input_)
  time_part1 = round((time.time()-t0)*1e3)
  print(f'Solution to part one: {part1} (time taken {time_part1}[ms])')

  t0 = time.time()
  part2 = solve_part2(input_)
  time_part2 = round((time.time()-t0)*1e3)
  print(f'Solution to part one: {part2} (time taken {time_part2}[ms])')


if __name__ == '__main__':
  main()

@pytest.mark.parametrize("input1,output1", [
  ('../data/test_input0.txt', 2)
])

def test1(input1,output1):
  assert solve_part1(read_input(input1)) == output1

@pytest.mark.parametrize("input2,output2", [
  ('../data/test_input0.txt', 1)
])

def test1(input2,output2):
  assert solve_part2(read_input(input2)) == output2

File: 2020/day2/cpp/src/day2.cpp
#include "day2.hpp"

#include <algorithm>

#include "utils.hpp"

using day2::input_type;

input_type day2::transform_input(std::vector<std::string> &input) {
  input_type output;

  for (auto &inp : input) {
    std::vector<std::string> split_input;

    inp.erase(std::remove(inp.begin(), inp.end(), ':'), inp.end());
    utils::split_string(inp, ' ', split_input);
    auto count = split_input[0], letter = split_input[1], pw = split_input[2];

    std::vector<std::string> split_count;
    utils::split_string(count, '-', split_count);
    int low = std::stoi(split_count[0]), high = std::stoi(split_count[1]);

    output.push_back(std::make_tuple(low, high, letter[0], pw));
  }

  return output;
}

input_type day2::read_input(std::string file_name) {
  auto input = utils::read_file(file_name);
  return transform_input(input);
}

// Implement part one solution
std::string day2::solve_part1(const input_type &input_) {
  int n_pws(0);
  for (auto &[low, high, letter, pw] : input_) {
    auto n_matches = std::count(pw.begin(), pw.end(), letter);
    n_pws += (low <= n_matches && n_matches <= high) ? 1 : 0;
  }
  return std::to_string(n_pws);
}

// Implement part two solution
std::string day2::solve_part2(const input_type &input_) {
  int n_pws(0);
  for (auto &i : input_) {
    auto low = std::get<0>(i), high = std::get<1>(i);
    auto letter = std::get<2>(i);
    auto pw = std::get<3>(i);

    n_pws += (pw[low - 1] == letter) ^ (pw[high - 1] == letter) ? 1 : 0;
  }
  return std::to_string(n_pws);
}
